core: Parse `module` correctly when `allow-unregistered-dialect` is set

alexarice opened the pull request:
Fixes #3668

dshaaban01 reviewed (approved):
Amazing🥳 thank you!

superlopuh reviewed (commented):
This LGTM, but I'm wondering whether we could simplify things by passing the dialect stack to `get_optional_op` directly, avoiding the creation of extra ops?

alexarice commented:
> This LGTM, but I'm wondering whether we could simplify things by passing the dialect stack to `get_optional_op` directly, avoiding the creation of extra ops?

Sounds good, I knew there was something I didn't like about my solution

alexarice commented:
I've rewritten it with @superlopuh suggestion. This didn't quite fix the needless creation of unregistered ops so I've moved some things to a helper function. Also added 2 unit tests.

math-fehr reviewed (approved):
Nice!

superlopuh commented on tests/test_mlcontext.py:
```diff
@@ -60,6 +60,40 @@ def test_get_op_unregistered():
+
+    assert ctx.get_optional_op("dummy", dialect_stack=("test",)) == DummyOp
+    op_type = ctx.get_optional_op("dummy2", dialect_stack=("test",))
+    print(op_type)
```
Could you please open a mini PR removing this?

alexarice replied:
#3673
